site/app/controllers/grading/ElectronicGraderController.php

```diff
@@ -59,13 +59,48 @@ public function run() {
             case 'verify_all':
                 $this->verifyGrader(true);
                 break;
+			case 'remove_empty':
+				return $this->ajaxRemoveEmpty();
+				break;
             default:
                 $this->showStatus();
                 break;
         }
     }
-    private function verifyGrader($verifyAll = false){
 
+    public function ajaxRemoveEmpty(){
+        //This function shows the empty spaces in the diffViewer
+        //TODO: Need to add checks?
+        $gradeable_id = $_REQUEST['gradeable_id'];
+        $gradeable = $this->core->getQueries()->getGradeable($gradeable_id, $_REQUEST['who_id']);
+        $gradeable->loadResultDetails();
+        $testcase = $gradeable->getTestcases()[$_REQUEST['index']];
+        //There are three options: original (Don't show empty space), escape (with escape codes), and unicode (with characters)
+        $option = isset($_REQUEST['option']) ? $_REQUEST['option'] : 'original';
+        //There are currently two views, the view of student's code and the expected view.
+        $which = isset($_REQUEST['which']) ? $_REQUEST['which'] : 'actual';
+        $autocheck_cnt = isset($_REQUEST['autocheck_cnt'])  ? intval($_REQUEST['autocheck_cnt']) : 0;
+        foreach ($testcase->getAutochecks() as $autocheck) {
+            $diff_viewer = $autocheck->getDiffViewer();
+            if($autocheck_cnt <= 0) {
+                break;
+            }
+            $autocheck_cnt -= 1;
+        }
+        $this->core->getOutput()->useFooter(false);
+        $this->core->getOutput()->useHeader(false);
+        $html = "";
+        if($which == "actual"){
+            $html .= $diff_viewer->getDisplayActual($option);
+        }
+        else {
+            $html .= $diff_viewer->getDisplayExpected($option);
+        }
+        $white_spaces = $diff_viewer->getWhiteSpaces();
+        echo json_encode(array("html"=>"$html", "whitespaces"=>$white_spaces));
```

Comment:
you can get rid of lines 90-91 and the echo and just do:
`$this->core->renderJson(['html' => $html, 'whitespaces' => $white_spaces]);`